plugins/snowflake/dbt/include/snowflake/macros/materializations/incremental.sql

```diff
@@ -42,8 +45,15 @@
 
     {%- call statement('main') -%}
       {{ create_table_as(false, target_relation, sql) }}
+
+      {%- if cluster_by_keys is not none -%}
```

Comment:
I don't think we need to alter the table clustering here -- the `snowflake__create_table_as` implementation should handle that part for us!

Reply:
Instead, we might need to add another config to enable/disable automatic clustering on the table. If incremental models are configured with automatic reclustering, then any dml statements made against the table (eg. merge/insert/delete) will trigger Snowflake to recluster the table.

The table should be in a mostly-clustered state because we've initialized it using an `order by` statement in `snowflake__create_table_as`, so ideally, Snowflake would just recluster a few of the (potentially very many) partitions in the incremental model.

Reply:
ok hmmm, so one thing that I may have failed to understand in the discussion of the issue was the effect of the `order by` clause. 
I get that on table creation order by will essentially cluster the table physically as it should be. But as far as I can understand it, this does not make snowflake aware that on DML actions like updates or inserts it may have to consider doing a bit of reclustering, especially in the case where the table is clustered say by `user_id` and `modified_at` kind of strategies. Although, arguably this kind of clustering may not be super useful. I only really see clustering being useful on tables like events table where actually you'd just cluster things once by an order by and since your incremental loads are always monotonic a you'd get the same benefits as having automatic clustering.

Reply:
Oh, sure, I think my response above was a little confusing.

I'd just call `alter table {{relation}} cluster by ({{cluster_by_keys}})` in the `snowflake__create_table_as` macro body, not in the materialization code. This means that every dbt materialization (including snapshots, for instance) can take advantage of clustering keys.

Reply:
Hehe that's actually what I did originally, but for some reason, there might be some machinery in place in your behind the scenes that escaped the user SQL at the end of the `create or replace ... as ( ... );` because the `alter` statement that I stuck after never got compiled. Could that be the case? Or maybe it was a silly PEBKAC error in one of my if statements or so...

Reply:
was a stupid mistake that I did not spot for a while...